sources/file.go

```diff
@@ -0,0 +1,112 @@
+package sources
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"path/filepath"
+	"strings"
+
+	"github.com/h2non/filetype"
+
+	"github.com/zricethezav/gitleaks/v8/logging"
+)
+
+// File implements Source for scanning a reader with a path
+type File struct {
+	// Content provides a reader to the file's content
+	Content io.Reader
+	// Path is the resolved real path of the file
+	Path string
+	// Symlink represents a symlink to the file if that's how it was discovered
+	Symlink string
+	// ChunkSize allows you to set the size of the chunks processed for this file
+	// the default is the 'chunkSize' constant in the sources.go file
+	ChunkSize int
+}
+
+func (s *File) Fragments(yield FragmentsFunc) error {
+	if s.ChunkSize == 0 {
+		s.ChunkSize = chunkSize
+	}
+
+	var (
+		// Buffer to hold file chunks
+		reader     = bufio.NewReaderSize(s.Content, s.ChunkSize)
+		buf        = make([]byte, s.ChunkSize)
+		totalLines = 0
+	)
+
+	for {
+		n, err := reader.Read(buf)
+		if n == 0 {
+			if err != nil && err != io.EOF {
+				return yield(Fragment{}, fmt.Errorf("could not read file: %w", err))
+			}
+
+			return nil
+		}
```

Comment:
IMO it's more straightforward to do `if n > 0`, rather than checking the error in two different places.

Reply:
I kept changing it around to do that but also not pass EOF to yield && still passing any amount of fragment and each way felt awkward to me and ended up being about the same amount of checks. 

I could also just be missing a simple way to do it though.

Reply:
Hmm, I will try to play around with this and see if there's a better way.